Saycode refactor, unit tests, and fixes

timothymtorres opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

Fixes #69798
Fixes #71621

When using hypnosis on a victim, the language should be accounted for and whether the victim can properly hear it.  Before the hearing code would magically translate any message, this is no longer the case.

This also fixes the language barrier involving hearing for:
- Mind echo trauma
- Phobia trauma
- Hypnotic trigger trauma
- Split Personality brainwashing trauma
- Codeword hearing
- Hypnotize status effect
- Impure Inacusiate reagent

## Why It's Good For The Game

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

Better consistency, improved readability, and less bugs in the future.

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
fix: Fix hypnosis, mind echo trauma, phobia trauma, hypnotic trigger trauma, split personality brainwashing trauma, codeword hearing, and impure inacusiate reagent all bypassing language and hearing checks.  If you try to give commands to a victim in a language they don't understand, they will no longer magically understand the words.
fix: Fix sign language having accent modifications
refactor: Refactored saycode to be more robust, readable, and have more unit tests.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

timothymtorres committed:
Fix hypnosis bypassing language barriers

MLGTASTICa commented:
ok but how are they meant to translate it if they find a speaker of the language its said in ? considering the limitations of the speech system ,this just seems to turn the item more niche.(Especially since there is no concrete way to translate from other languages, with garbled input)

timothymtorres commented:
So this is the code that is used to hear messages and receive a translation:

https://github.com/tgstation/tgstation/blob/f1a363c825a31b9083019d19c69212e87fd78722/code/game/say.dm#L35-L80

https://github.com/tgstation/tgstation/blob/c662f6791a8e893755c0988cc796fc9a72c5a340/code/__DEFINES/dcs/signals/signals_atom/signals_atom_movable.dm#L52-L60

https://github.com/tgstation/tgstation/blob/e0590c3f5bcffe844c18c33c09c1a96f2acd1fc2/code/datums/status_effects/debuffs/debuffs.dm#L556-L559

The garbled speech is like **Lorem Ipsum** and is randomized.  The whole point of this is to nullify messages that a hearer can't understand like in the issue report.

timothymtorres committed:
Standardize hearing checks

timothymtorres committed:
Fix owner_hear arg for inacusiate

timothymtorres commented:
Tested and ready!

ATH1909 commented:
okay but being hypnotized by the phrase "ook ook eek" is funny

SuperNovaa41 commented:
ATH is right, I think it would be funny to be hypnotized by monkey speak. Just make it not translate.

timothymtorres committed:
Refactor speech code to allow translation arg

timothymtorres committed:
Refactor language translation code

timothymtorres committed:
Remove deprecated say code

timothymtorres committed:
Fix messages not appearing

timothymtorres committed:
Remove deprecated code

timothymtorres committed:
Remove deprecated code comments

timothymtorres committed:
Add language unit test

timothymtorres committed:
Remove more deprecated speech code comments

timothymtorres committed:
Add spacing to beetlejuice args

timothymtorres committed:
Add unit tests for say code

timothymtorres committed:
Merge remote-tracking branch 'upstream/master' into patch-13

Mothblocks commented on code/modules/unit_tests/say.dm:
```diff
@@ -40,3 +40,92 @@
+	host_mob.grant_language(/datum/language/beachbum, spoken=TRUE, understood=TRUE) // can now understand
+	TEST_ASSERT_EQUAL(surfer_quote, host_mob.translate_language(host_mob, /datum/language/beachbum, surfer_quote), "Language test failed. Mob was supposed NOT to understand: [surfer_quote]")
+
+/datum/unit_test/speech
```
Document the test

timothymtorres replied:
Done with https://github.com/tgstation/tgstation/pull/69799/commits/c2424862585f3e2ab030c2d8e7c749959c2cc4de

timothymtorres committed:
Add more unit tests for speech

timothymtorres commented on code/modules/unit_tests/say.dm:
```diff
@@ -40,3 +40,92 @@
 /mob/living
 	var/last_say_args_ref
+
+/datum/unit_test/translate_language
```
Done with https://github.com/tgstation/tgstation/pull/69799/commits/c2424862585f3e2ab030c2d8e7c749959c2cc4de

timothymtorres committed:
Remove deprecated code comment

timothymtorres committed:
Remove owner needing language to trigger effects

Mothblocks commented on code/modules/unit_tests/say.dm:
```diff
@@ -41,6 +41,7 @@
 /mob/living
 	var/last_say_args_ref
 
+/// This unit test trnanslates a string from one language to another depending on if the person can understand the
```
```suggestion
/// This unit test translates a string from one language to another depending on if the person can understand the
```

Mothblocks commented on code/modules/unit_tests/say.dm:
```diff
@@ -112,20 +104,54 @@
 	TEST_ASSERT(handle_speech_result, "Handle speech signal was not fired")
 	TEST_ASSERT(handle_hearing_result, "Handle hearing signal was not fired")
+	handle_speech_result = FALSE
+	handle_hearing_result = FALSE
```
Instead of doing this, just make these a counter that ticks up, and check the times it was called.

timothymtorres replied:
I didn't like how I was c/p'ing the same code repeatedly so I refactored it into a proc that resets those vars properly.  Look at https://github.com/tgstation/tgstation/pull/69799/commits/970a637fe63743b01e405073e1be9581fa2be20c to see how ugly it was looking.  I think it's way more readable now and allows for more testing in the future.

timothymtorres committed:
Fix syntax error

timothymtorres committed:
Fix spelling for code comment

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

timothymtorres committed:
Merge remote-tracking branch 'upstream/master' into patch-13

timothymtorres committed:
Add incremental counter for say unit test

timothymtorres committed:
Fix syntax error for sign langauge

Mothblocks commented on code/modules/unit_tests/say.dm:
```diff
@@ -110,45 +110,33 @@
-	TEST_ASSERT(handle_hearing_result, "Handle hearing signal was not fired")
-	handle_speech_result = FALSE
-	handle_hearing_result = FALSE
+	TEST_ASSERT(handle_speech_result == 1, "Handle speech signal was not fired")
```
TEST_ASSERT_EQUAL for this and everything else

timothymtorres replied:
https://github.com/tgstation/tgstation/blob/ad3aea8f6f2778d8c1fabd8cf95bfcb0f56c5bda/code/modules/unit_tests/say.dm#L108-L124

timothymtorres committed:
Refactor unit test asserts to be less verbose

timothymtorres committed:
Fix incorrect var name

timothymtorres committed:
To be continued...

timothymtorres committed:
Refactor speech unit tests to be sane

timothymtorres committed:
Fix variable list types

timothymtorres committed:
Fix syntax error

timothymtorres committed:
Fix mob scope

timothymtorres committed:
3rd times the charm?

timothymtorres committed:
Merge branch 'master' into patch-13

timothymtorres committed:
Add range speech arg for unit tests

timothymtorres committed:
Merge remote-tracking branch 'upstream/master' into patch-13

timothymtorres committed:
Merge remote-tracking branch 'upstream/master' into patch-13

Mothblocks commented:
TMing on Manuel

timothymtorres committed:
Fix listeners not translating languages properly

timothymtorres committed:
Fix listener to hear species speech modifier

timothymtorres committed:
Refactor saycode to be EVEN MORE POWERFUL

timothymtorres committed:
Add extra unit tests for language

timothymtorres committed:
Merge remote-tracking branch 'upstream/master' into patch-13

timothymtorres committed:
Fix .proc from say unit tests

timothymtorres committed:
Add spacing to commas

timothymtorres committed:
Fix runechat not processing languages or whispers

timothymtorres committed:
Fix speech unit tests language args

timothymtorres committed:
Comment out translation code for runechat

timothymtorres committed:
Refactor deaf_type to use defines

timothymtorres committed:
Fix whispering and translation to apply to sign language

timothymtorres committed:
Fix say unit test failure

timothymtorres committed:
Fix say unit test to set lists to null at startup

timothymtorres committed:
Fix listener not in the proper position

timothymtorres committed:
Another attempt at fixing listener position

timothymtorres committed:
Fix unit test failing due to missing client

timothymtorres committed:
To be continued

timothymtorres committed:
Fix language unit test to work

timothymtorres committed:
Remove deprecated say code

timothymtorres committed:
Merge remote-tracking branch 'upstream/master' into patch-13

timothymtorres committed:
Fix malformed message arg

timothymtorres committed:
Remove deprecated sign language treat message signal

No need to have this.  It can be put into COMSIG_MOB_SAY and remove the other signal.

timothymtorres committed:
Merge remote-tracking branch 'upstream/master' into patch-13

timothymtorres commented:
This was test merged for a decent amount of time with no reported problems.  Should be okay to merge now.

timothymtorres committed:
Fix sign language having accent modifications

timothymtorres committed:
Fix wrong syntax name
